pkg/roachprod/roachprodutil/codec/types.go

```diff
@@ -0,0 +1,201 @@
+// must be registered before they can be encoded and decoded.
+func Register(t DynamicType) {
+	typeRegistry[t.GetTypeName()] = ResolveElem(reflect.TypeOf(t))
+}
+
+// Wrap is a convenience function for wrapping a dynamic type.
+func Wrap[T DynamicType](t T) Wrapper[T] {
+	return Wrapper[T]{Val: t}
+}
+
+// WrapList is a convenience function for wrapping a list of dynamic types.
+func WrapList[T DynamicType](t []T) ListWrapper[T] {
+	return ListWrapper[T]{Val: t}
+}
+
+func (w *Wrapper[T]) Get() T {
+	return w.Val
+}
+
+func (w *ListWrapper[T]) Get() []T {
+	return w.Val
+}
+
+// ResolveElem unwraps a pointer type to get the underlying type or returns the
+// type itself if it is not a pointer.
+func ResolveElem(t reflect.Type) reflect.Type {
+	if t.Kind() == reflect.Ptr {
+		return t.Elem()
+	}
+	return t
+}
+
+// ResolveTypeName resolves the type name of a dynamic type.
+func ResolveTypeName(t DynamicType) TypeName {
+	tType := ResolveElem(reflect.TypeOf(t))
+	return TypeName{
+		pkg:  path.Base(tType.PkgPath()),
+		name: tType.Name(),
+	}
+}
+
+// toFieldType encodes the type name as a fieldType (string), including the
+// package path, and if the type should later on be decoded as a pointer.
+func (t TypeName) toFieldType(pointer bool) fieldType {
+	name := t.name
+	if pointer {
+		name = "*" + name
+	}
+	return fieldType(t.pkg + "." + name)
+}
+
+// name returns the package and name of the type.
+func (t fieldType) name() (TypeName, error) {
+	s := strings.Split(string(t), ".")
+	if len(s) != 2 {
+		return TypeName{}, fmt.Errorf("invalid type name %q", t)
+	}
+	// The `typeName` should not include the pointer prefix, as this is only used
+	// during decoding to determine if the value should be decoded as a pointer.
+	s[1] = strings.TrimPrefix(s[1], "*")
+	return TypeName{
+		pkg:  s[0],
+		name: s[1],
+	}, nil
+}
+
+// isPointer returns true if the type is a pointer.
+func (t fieldType) isPointer() bool {
+	return strings.Contains(string(t), "*")
+}
+
+// MarshalYAML implements the [yaml.Marshaler] interface. The wrapper is
+// inspected to get its type information, which is then encoded with the
+// metadata type.
+func (w Wrapper[T]) MarshalYAML() (any, error) {
+	typeName := w.Get().GetTypeName()
+	dynType := reflect.TypeOf(w.Get())
+	pointer := false
+	if dynType.Kind() == reflect.Ptr {
+		pointer = true
+	}
+	return metadata{
+		FieldType: typeName.toFieldType(pointer),
+		Val:       w.Val,
+	}, nil
+}
+
+// UnmarshalYAML implements the [yaml.Unmarshaler] interface. Since the wrapper
+// was stored with metadata, we decode the type information and then decode the
+// value using the type information.
+func (w *Wrapper[T]) UnmarshalYAML(value *yaml.Node) error {
+	ft := fieldType(value.Content[1].Value)
+	typeName, err := ft.name()
+	if err != nil {
+		return err
+	}
+	objType, ok := typeRegistry[typeName]
+	if !ok {
+		return fmt.Errorf("unknown type %s.%s", typeName.pkg, typeName.name)
```

Comment:
I wonder if this should return an error or default to the generic `map[string]interface{}` container when the type is unknown.

I can see a situation in which we're trying to use an older version to parse a plan generated by a more recent version that embeds types unknown to the unmarshaller. I think there is a case to be made for forward compatibility in the decoder and warns/errors in the code that uses the strucs afterwards.